File: matlive-sdk/src/main/java/com/matnsolutions/matlive_sdk/audio/mangers/mat_live_join_room_manger.dart
import 'dart:async';

import 'package:livekit_client/livekit_client.dart';
import 'package:matlive_client_sdk_flutter/audio/define/mat_live_user.dart';
import 'package:matlive_client_sdk_flutter/audio/mangers/mat_live_room_manger.dart';
import 'package:matlive_client_sdk_flutter/services/utlis.dart';
import 'package:matlive_client_sdk_flutter/utils/kprint.dart';
import 'package:permission_handler/permission_handler.dart';

class JoinRequest {
  JoinRequest({
    this.url = "",
    this.token = "",
    this.e2ee = false,
    this.e2eeKey,
    this.simulcast = true,
    this.adaptiveStream = true,
    this.dynacast = true,
    this.preferredCodec = 'VP8',
    this.enableBackupVideoCodec = true,
  });

  String url;
  String token;
  final bool e2ee;
  final String? e2eeKey;
  final bool simulcast;
  final bool adaptiveStream;
  final bool dynacast;
  final String preferredCodec;
  final bool enableBackupVideoCodec;
}

class MatLiveJoinRoomManger {
  static final MatLiveJoinRoomManger instance =
      MatLiveJoinRoomManger._internal();

  MatLiveJoinRoomManger._internal();

  final JoinRequest _request = JoinRequest();
  List<MediaDevice> _audioInputs = [];
  StreamSubscription? _subscription;
  MatLiveUser? currentUser;
  LocalAudioTrack? audioTrack;
  String roomId = "";
  MediaDevice? _selectedAudioDevice;
  Function(int seatIndex)? onInvitedToMic;
  Function(String data)? onSendGift;

  Future<void> init({
    required Function(int seatIndex)? onInvitedToMic,
    required Function(String data)? onSendGift,
  }) async {
    _request.url = Utlis.url;
    this.onInvitedToMic = onInvitedToMic;
    this.onSendGift = onSendGift;
    if (lkPlatformIs(PlatformType.android)) {
      await _checkPermissions();
    }
    if (lkPlatformIsMobile()) {
      await LiveKitClient.initialize(bypassVoiceProcessing: true);
    }
    _subscription =
        Hardware.instance.onDeviceChange.stream.listen(_loadDevices);
    await Hardware.instance.enumerateDevices().then(_loadDevices);
  }

  Future<void> _checkPermissions() async {
    final res = await Future.wait([
      Permission.bluetooth.request(),
      Permission.bluetoothConnect.request(),
      Permission.microphone.request(),
    ]);
    kPrint(res);
  }

  Future<void> close() async {
    _subscription?.cancel();
    await _stopAudioStream();
    await MatLiveRoomManger.instance.close();
  }

  Future<void> _loadDevices(List<MediaDevice> devices) async {
    _audioInputs = devices.where((d) => d.kind == 'audioinput').toList();

    if (_audioInputs.isNotEmpty) {
      if (_selectedAudioDevice == null) {
        _selectedAudioDevice = _audioInputs.first;
        await _changeLocalAudioTrack();
      }
    }
  }

  Future<void> _stopAudioStream() async {
    await audioTrack?.stop();
    audioTrack = null;
  }

  Future<void> _changeLocalAudioTrack() async {
    if (audioTrack != null) {
      await audioTrack!.stop();
      audioTrack = null;
    }

    if (_selectedAudioDevice != null) {
      audioTrack = await LocalAudioTrack.create(AudioCaptureOptions(
        deviceId: _selectedAudioDevice!.deviceId,
      ));
    }
  }

  Future<void> connect({
    required String token,
    required String name,
    required String avatar,
    required String userId,
    required String roomId,
    String? metadata,
  }) async {
    this.roomId = roomId;
    _request.token = token;
    currentUser = MatLiveUser(
      name: name,
      avatar: avatar,
      userId: userId,
      roomId: userId,
      metadata: metadata,
    );
    try {
      E2EEOptions? e2eeOptions;
      if (_request.e2ee && _request.e2eeKey != null) {
        final keyProvider = await BaseKeyProvider.create();
        e2eeOptions = E2EEOptions(keyProvider: keyProvider);
        await keyProvider.setKey(_request.e2eeKey!);
      }

      final room = Room(
        roomOptions: RoomOptions(
          adaptiveStream: _request.adaptiveStream,
          dynacast: _request.dynacast,
          defaultAudioPublishOptions: const AudioPublishOptions(
            name: 'custom_audio_track_name',
            audioBitrate: AudioPreset.musicHighQualityStereo,
          ),
          e2eeOptions: e2eeOptions,
        ),
      );
      await room.prepareConnection(_request.url, _request.token);
      await room.connect(
        _request.url,
        _request.token,
        fastConnectOptions: FastConnectOptions(
          microphone: TrackOption(track: audioTrack),
        ),
      );
      MatLiveRoomManger.instance.room = room;
      MatLiveRoomManger.instance.listener = room.createListener();
      MatLiveRoomManger.instance.setUp(
        onInvitedToMic,
        onSendGift,
      );
    } catch (error) {
      kPrint('Could not connect $error');
    }
  }
}

File: matlive-sdk/src/main/java/com/matnsolutions/matlive_sdk/audio/mangers/mat_live_join_room_manger.kt
package com.matnsolutions.matlive_sdk.audio.mangers

import Utlis
import android.content.Context
import com.matnsolutions.matlive_sdk.audio.define.MatLiveUser
import com.matnsolutions.matlive_sdk.utils.kPrint
import io.livekit.android.AudioOptions
import io.livekit.android.LiveKit
import io.livekit.android.LiveKitOverrides
import io.livekit.android.RoomOptions
import io.livekit.android.audio.AudioProcessorOptions
import io.livekit.android.audio.AudioSwitchHandler
import io.livekit.android.e2ee.BaseKeyProvider
import io.livekit.android.e2ee.E2EEOptions
import io.livekit.android.room.participant.AudioTrackPublishDefaults
import io.livekit.android.room.track.LocalAudioTrack
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

data class JoinRequest(
    var url: String = "",
    var token: String = "",
    val e2ee: Boolean = false,
    val e2eeKey: String? = null,
    val simulcast: Boolean = true,
    val adaptiveStream: Boolean = true,
    val dynacast: Boolean = true,
    val preferredCodec: String = "VP8",
    val enableBackupVideoCodec: Boolean = true,
)

class MatLiveJoinRoomManger private constructor() {
    companion object {
        val instance: MatLiveJoinRoomManger by lazy { MatLiveJoinRoomManger() }
    }

    private val _request = JoinRequest()

    var currentUser: MatLiveUser? = null
    var audioTrack: LocalAudioTrack? = null
    var roomId: String = ""

    var onInvitedToMic: ((Int) -> Unit)? = null
    var onSendGift: ((String) -> Unit)? = null
    private val coroutineScope = CoroutineScope(Dispatchers.IO)

    fun init(
        onInvitedToMic: ((Int) -> Unit)?,
        onSendGift: ((String) -> Unit)?,
    ) {
        _request.url = Utlis.url
        this.onInvitedToMic = onInvitedToMic
        this.onSendGift = onSendGift
//        initLocalAudioTrack();
    }


//    suspend fun initLocalAudioTrack() {
//        audioTrack = LocalAudioTrack(
//            AudioCaptureOptions(
//                noiseSuppression = true, echoCancellation = true
//            )
//        )
//    }

    suspend fun close() {
        _stopAudioStream()
        MatLiveRoomManger.instance.close()
    }


    private suspend fun _stopAudioStream() {
        audioTrack?.stop()
        audioTrack = null
    }


    //    private  lateinit var application: Application;
    suspend fun connect(
        context: Context,
        token: String,
        name: String,
        avatar: String,
        userId: String,
        roomId: String,
        metadata: String? = null,
    ) {
        this.roomId = roomId
//        this.application = application
        _request.token = token
        currentUser = MatLiveUser(
            name = name,
            avatar = avatar,
            userId = userId,
            roomId = userId,
            metadata = metadata,
        )
        try {
            var e2eeOptions: E2EEOptions? = null
            if (_request.e2ee && _request.e2eeKey != null) {
                val keyProvider = BaseKeyProvider()
                e2eeOptions = E2EEOptions(keyProvider = keyProvider)
                keyProvider.setKey(key = _request.e2eeKey, participantId = userId)
            }

            val room = LiveKit.create(
                appContext = context,
                options = RoomOptions(
                    adaptiveStream = _request.adaptiveStream,
                    dynacast = _request.dynacast,
                    e2eeOptions = e2eeOptions,
//                    audioTrackPublishDefaults = AudioTrackPublishDefaults(audioBitrate = 3200)
                ),
                overrides = LiveKitOverrides(
                    audioOptions = AudioOptions(audioProcessorOptions = AudioProcessorOptions()),
                ),
            )
            room.prepareConnection(_request.url, _request.token)

            coroutineScope.launch {
                room.connect(
                    _request.url,
                    _request.token,
//                options = ConnectOptions(
//                    microphone = TrackOption(track = audioTrack),
//                )
                )
            }
            // Start a foreground service to keep the call from being interrupted if the
            // app goes into the background.
//            val foregroundServiceIntent = Intent(application, ForegroundService::class.java)
//            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
//                application.startForegroundService(foregroundServiceIntent)
//            } else {
//                application.startService(foregroundServiceIntent)
//            }
            MatLiveRoomManger.instance.room = room
//            MatLiveRoomManger.instance.listener = room.createListener()
            MatLiveRoomManger.instance.setUp(
                onInvitedToMic,
                onSendGift,
            )
            delay(1000)
            val audioHandler = room.audioHandler as AudioSwitchHandler
            audioHandler.let {
                it.selectDevice(it.availableAudioDevices.last())
            }
        } catch (error: Exception) {
            kPrint("Could not connect $error")
            throw Exception("Failed to update metadata: $error");
        }
    }
}
